retrieval/target.go

```diff
@@ -267,6 +266,30 @@ func (t *Target) String() string {
 	return t.host()
 }
 
+// fingerprint returns an identifying hash for the target.
+func (t *Target) fingerprint() model.Fingerprint {
+	t.RLock()
+	defer t.RUnlock()
+
+	return t.labels.Fingerprint()
+}
+
+// offset returns the time until the next scrape cycle for the target.
+func (t *Target) offset(interval time.Duration) time.Duration {
+	now := time.Now().UnixNano()
+
+	var (
+		base   = now % int64(interval)
```

Comment:
That's essentially the base offset. To that we add another offeset between 0 and `interval` based on the fingerprint.
Finally, if this total offest (`next`) falls behind now, we move it one cycle ahead.

The returned duration is the wait time, not the absolute time of the next scrape.